File: Spring_Study/spring-security/src/main/java/com/example/springsecurity/repository/UserRepository.java
package com.example.springsecurity.repository;

import com.example.springsecurity.model.User;
import org.springframework.data.jpa.repository.JpaRepository;

// CRUD 함수를 JpaRepository가 갖고 있음
// @Repository라는 어노테이션이 없어도 IoC가 됨 => JpaRepository를 상속했기 때문
public interface UserRepository extends JpaRepository<User, Integer> {
    // select * from user where username = 1?
    public User findByUsername(String username);

    // select * from user where email = 1?
    public User findByEmail(String email);

}

File: Kotlin_Study/library-app/src/main/kotlin/com/group/libraryapp/domain/user/UserRepository.kt
package com.group.libraryapp.domain.user

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.util.Optional

interface UserRepository : JpaRepository<User, Long> {
    fun findByName(name: String): User? // Optional<User>
    // Optional은 어떤 값이 Null인지 아닌지를 나타내기 위해서 JDK8부터 등장했지만,
    // 코틀린에서는 타입?를 통해서 이를 해결할 수 있으므로 크게 필요가 없다.

    @Query("SELECT DISTINCT u FROM User u LEFT JOIN FETCH u.userLoanHistories") // 유저 테이블과 (유저의) 유저 히스토리 테이블을 left join함
    fun findAllWithHistories(): List<User>

}